synapse/rest/client/v1/login.py

```diff
@@ -298,11 +301,19 @@ async def _do_jwt_login(self, login_submission: JsonDict) -> Dict[str, str]:
             )
 
         import jwt
+        from jwt import PyJWKClient
+
+        key = self.jwt_secret
+
+        if not key and self.jwt_jwks_uri:
+            jwks_client = PyJWKClient(self.jwt_jwks_uri)
+            signing_key = jwks_client.get_signing_key_from_jwt(token)
+            key = signing_key.key
```

Comment:
It would be ideal if we only had to fetch this once (similar to the `_provider_metadata` property of `OidcHandler`).